go-controller/vendor/github.com/coreos/go-iptables/iptables/iptables.go

```diff
@@ -185,6 +220,43 @@ func (ipt *IPTables) Insert(table, chain string, pos int, rulespec ...string) er
 	return ipt.run(cmd...)
 }
 
+// Replace replaces rulespec to specified table/chain (in specified pos)
+func (ipt *IPTables) Replace(table, chain string, pos int, rulespec ...string) error {
+	cmd := append([]string{"-t", table, "-R", chain, strconv.Itoa(pos)}, rulespec...)
+	return ipt.run(cmd...)
+}
+
+// InsertUnique acts like Insert except that it won't insert a duplicate (no matter the position in the chain)
+func (ipt *IPTables) InsertUnique(table, chain string, pos int, rulespec ...string) error {
+	exists, err := ipt.Exists(table, chain, rulespec...)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return ipt.Insert(table, chain, pos, rulespec...)
+	}
+
+	return nil
+}
+
+// Restore replaces specified chains and rules in a specific table
+// rulesMap is keyed by chain name, and holds slices of rulespecs
+// Only chains specified in the map will be flushed and replaced. Other chains will not be affected.
+func (ipt *IPTables) Restore(table string, rulesMap map[string][][]string) error {
+	restoreRules := "*" + table
+	for chain, rules := range rulesMap {
+		restoreRules += "\n" + fmt.Sprintf(":%s - [0:0]", strings.ToUpper(chain))
+		for _, rule := range rules {
+			restoreRules += "\n" + fmt.Sprintf("-I %s %s", chain, strings.Join(rule, " "))
```

Comment:
restore is inserting the rules? so the last rule in the map is what will have the highest precedence? so users have to be careful to order the rules in a specific fashion if they need to.

Reply:
yeah thats a bit of a subtle trick I did to keep the order consistent with what OVNK did before. I think semantically it probably makes more sense to make it all append, and I dont think the order really matters for most of our rules...but i didnt want to risk breaking anything so I kept it the same.